File: cs25backend/src/main/java/edu/dongguk/cs25backend/domain/ItemCS.java
package edu.dongguk.cs25backend.domain;

import edu.dongguk.cs25backend.domain.type.ItemCategory;
import edu.dongguk.cs25backend.exception.CS25Exception;
import edu.dongguk.cs25backend.exception.ErrorCode;
import jakarta.persistence.*;
import lombok.Builder;
import lombok.Getter;
import lombok.NoArgsConstructor;
import org.hibernate.annotations.DynamicUpdate;

import java.util.ArrayList;
import java.util.List;

import static jakarta.persistence.CascadeType.ALL;

@Entity
@Getter
@NoArgsConstructor
@DynamicUpdate
@Table(name = "item_CS")
public class ItemCS {
    @Id
    @GeneratedValue(strategy = GenerationType.IDENTITY)
    @Column(name = "item_cs_id")
    private Long id;

    private String name;

    private int price;

    private int stock;

    private String thumbnail;

    @Enumerated(EnumType.STRING)
    private ItemCategory itemCategory;

    //== 연관 관계 매핑 ==//

    @ManyToOne(fetch = FetchType.LAZY)
    @JoinColumn(name = "item_hq_id")
    private ItemHQ itemHQ;

    @OneToMany(mappedBy = "itemCS", cascade = ALL)
    private List<OrderItemCS> orderItems = new ArrayList<>();

    @ManyToOne(fetch = FetchType.LAZY)
    @JoinColumn(name = "store_id")
    private Store store;


    public void addStock(int stock) {
        this.stock += stock;
    }

    public void removeStock(int stock) {
        int restStock = this.stock - stock;
        if (restStock < 0) {
            throw new CS25Exception(ErrorCode.NOT_ENOUGH_ERROR);
        }
        this.stock = restStock;
    }

    @Builder
    public ItemCS(String name, int price, int stock, String thumbnail, ItemCategory itemCategory, ItemHQ itemHQ, Store store) {
        this.name = name;
        this.price = price;
        this.stock = stock;
        this.thumbnail = thumbnail;
        this.itemCategory = itemCategory;
        this.itemHQ = itemHQ;
        this.store = store;
    }
}

File: cs25server/src/main/kotlin/edu/dongguk/cs25server/domain/ItemCS.kt
package edu.dongguk.cs25server.domain

import edu.dongguk.cs25server.domain.type.ItemCategory
import edu.dongguk.cs25server.dto.response.ItemsResponse
import edu.dongguk.cs25server.dto.response.StockForStoreDto
import edu.dongguk.cs25server.exception.GlobalException
import edu.dongguk.cs25server.exception.ErrorCode
import jakarta.persistence.*
import org.hibernate.annotations.DynamicUpdate

@Entity
@DynamicUpdate
@Table(name = "item_CS")
class ItemCS(
        var name: String,

        var price: Int = 0,

        var stock: Int = 0,

        var isNewItem: Boolean = true,

        @Enumerated(EnumType.STRING)
        var category: ItemCategory,

        @ManyToOne(fetch = FetchType.LAZY)
        @JoinColumn(name = "item_hq_id")
        var itemHQ: ItemHQ? = null,

        @ManyToOne(fetch = FetchType.LAZY)
        @JoinColumn(name = "store_id")
        var store: Store? = null,

        @ManyToOne(fetch = FetchType.LAZY)
        @JoinColumn(name = "image_id")
        var image: Image
) {
    @Id
    @GeneratedValue(strategy = GenerationType.IDENTITY)
    @Column(name = "item_cs_id")
    private val id: Long? = null

    @OneToMany(mappedBy = "itemCS", cascade = [CascadeType.ALL])
    private lateinit var orderItems: MutableList<OrderItemCS>

    fun getId() = this.id

    fun setStores(store: Store) {
        this.store = store
        store.getItemCS().add(this)
    }

    fun addStock(stock: Int) {
        this.stock += stock
    }

    fun checkStock(stock: Int) {
        if (this.stock - stock < 0) {
            throw GlobalException(ErrorCode.NOT_ENOUGH_STOCK_ERROR)
        }
    }

    fun removeStock(stock: Int) {
        val restStock = this.stock - stock
        if (restStock < 0) {
            throw GlobalException(ErrorCode.NOT_ENOUGH_STOCK_ERROR)
        }
        this.stock = restStock
    }

    fun toResponse(): StockForStoreDto = StockForStoreDto(
        id = this.id,
        name = this.name,
        amount = this.stock,
        price = this.price,
        category = this.category.toString()
    )

    fun toItemsResponse(): ItemsResponse {
        return ItemsResponse(
            id = this.id,
            name = this.name,
            price = this.price,
            thumbnail = this.image.getAccessUrl(),
            stock = this.stock
        )
    }

}


